src/conversion.c

```diff
@@ -299,15 +291,28 @@ int igraph_get_edgelist(const igraph_t *graph, igraph_vector_t *res, igraph_bool
 
     if (bycol) {
         while (!IGRAPH_EIT_END(edgeit)) {
-            igraph_edge(graph, IGRAPH_EIT_GET(edgeit), &from, &to);
+            long int edge = IGRAPH_EIT_GET(edgeit);
+            igraph_edge(graph, (igraph_integer_t) edge, &from, &to);
+	    /* FIXME: why are the two following lines not equivalent to igraph_edge??
```

Comment:
This one is really a mystery to me, @szhorvat @ntamas do you understand why the commented lines crash the tests?

Reply:
`igraph_edge()` does something weird: if the graph is directed, it works as expected (the result of `IGRAPH_FROM()` is placed in `from` and `IGRAPH_TO()` is placed in `to`), but if the graph is undirected, it does the _opposite_ (the result of `IGRAPH_TO()` goes in `from` and vice versa. I don't know why it is implemented this way, but the pattern is quite consistent in `type_indexededgelist.c`.

One thing that I figured out is that there are two vectors in `igraph_t` (`from` and `to`) that represent the source and target vertices of the edges in the graph. When igraph is adding an edge to an undirected graph, it always makes sure that the _larger_ vertex ID goes in `from` and the _smaller_ vertex ID goes in `to`. Assuming that the user's expectation is to see the smaller vertex ID first for an undirected edge, we need to swap `from` and `to` for undirected graphs.

Reply:
OMG.

@gaborcsardi can you perhaps remember why/how this came to be? For your convenience here the igraph_edge function which seems to do the "unexpected" thing for undirected graphs:

```C
int igraph_edge(const igraph_t *graph, igraph_integer_t eid,
                igraph_integer_t *from, igraph_integer_t *to) {

    if (igraph_is_directed(graph)) {
        *from = IGRAPH_FROM(graph, eid);
        *to   = IGRAPH_TO(graph, eid);
    } else {
        *from = IGRAPH_TO(graph, eid);
        *to   = IGRAPH_FROM(graph, eid);
    }

    return IGRAPH_SUCCESS;
}
```

@ntamas Rightio, for now I'll just triple check that all changes in this PR are consistent with the previous ordering. However, I am not sure everyone on the team (let away the users!) know this tricky thing and that the convention is:

**For undirected graphs, igraph stores the larger vertex id in `graph->from`, but `igraph_edge` and `igraph_edges` will swap the order to give the smaller vertex id first.**

I'll write a post on the forum just as a reminder.

Reply:
I am not sure what you mean specifically. It makes sense to always store the larger vertex first, because it makes operations simpler. Or always store the smaller first, it does not matter. I guess the choice was arbitrary, and then hard to change, because in R this data structure can be saved to disk.

I guess the `igraph_edge()` behavior came later, because it is indeed more natural for people to see the smaller vertex first.

Reply:
I see, thanks for that. I just didn't know that igraph_edge swaps things so it was surprising.

Reply:
> in R this data structure can be saved to disk.

I did not know that - we need to remember this in case we ever decide to change the graph representation. (I was under the impression that the internals of `igraph_t` are an implementation detail and we are free to change it if the basic interface does not change). It means that the best we can do if we change the representation is to provide _both_ the old and the new one, and let the user select it at compile time. The R interface could then stick to the old representation.

Reply:
You can still change it, just need to do more work when "converting" between `igraph_t` and the R object.

But why would you change it, anyway? There are _a lot_ of other possible improvements, with much bigger benefits than changing this, imo.